WalletWasabi.Fluent/Models/WebcamQrReader.cs

```diff
@@ -0,0 +1,166 @@
+using Avalonia.Media.Imaging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OpenCvSharp;
+using Avalonia;
+using Avalonia.Media;
+using System.Runtime.InteropServices;
+using WalletWasabi.Logging;
+using WalletWasabi.Userfacing;
+using NBitcoin;
+using System.Threading;
+
+namespace WalletWasabi.Fluent.Models
+{
+	public class WebcamQrReader
+	{
+		public bool RequestEnd { get; set; }
+		public Network Network { get; }
+		public Task? ScanningTask { get; set; }
+
+		public WebcamQrReader(Network network)
+		{
+			Network = network;
+		}
+
+		public void StartScanning()
+		{
+			if (ScanningTask is not { })
+			{
+				ScanningTask = Task.Run(() =>
+				{
+					VideoCapture? camera = null;
+					try
+					{
+						if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+						{
+							throw new NotImplementedException("This operating system is not supported.");
+						}
+						camera = OpenCamera();
+						RequestEnd = false;
+						Scan(camera);
+					}
+					catch (Exception exc)
+					{
+						Logger.LogError("QR scanning stopped. Reason:", exc);
+						ErrorOccured?.Invoke(this, exc);
+					}
+					finally
+					{
+						camera?.Release();
+					}
+				});
+			}
+		}
+
+		public async Task StopScanningAsync()
+		{
+			if (ScanningTask is { } task)
+			{
+				RequestEnd = true;
+				await task;
+
+				ScanningTask = null;
+			}
+		}
+
+		private void Scan(VideoCapture camera)
+		{
+			WriteableBitmap? lastBitmap = null;
+			WriteableBitmap? currentBitmap = null;
```

Comment:
Using of these two variable feels very suspicious to me.

Basically, an image is propagated to "SendViewModel" and once scanning stops (`RequestEnd` is true), you dispose the image on display on L115 and L116, if it makes sense what I'm saying. If true, that doesn't sound right to me.

Also if the Avalonia's `WriteableBitmap` is supposed to behave as https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.imaging.writeablebitmap?view=net-5.0 this one, then I believe you are supposed to create just one `WriteableBitmap` and update it while holding `map.Lock()`. Your solution IMHO leads to excessive garbage collection as you create new and new bitmaps and that most likely explains why on my computer the image changes are kind of slow.

Reply:
I would say that it's kind of slow because of the `ConvertMatToWriteableBitmap` function, which takes ~100ms to finish.
The version before this was to use one `WriteableBitmap` that we change, and dispose at the end, but that lead to the image to tick while reading as we disposed faster then we read the new image. When we just simply changed the bitmap to another new one, the last one didn't get to be disposed.
I will look into this if I can find a faster solution later on.

Reply:
Is there a `map.Lock()` function there? If yes you can try that and remove the extra logic.

Reply:
https://github.com/zkSNACKs/WalletWasabi/pull/6002/files#diff-429b41fcfab758aca9e607e12584630936d2cb37880c29afaf8d439ade4af302R143 ;-)